streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreSegmentValueFormatter.java

```diff
@@ -341,8 +345,10 @@ public void insertAsLatest(final long validFrom, final long validTo, final byte[
                 // detected inconsistency edge case where older segment has [a,b) while newer store
                 // has [a,c), due to [b,c) having failed to write to newer store.
                 // remove entries from this store until the overlap is resolved.
```

Comment:
Yep, that's spot on. I will update the comment with your suggestions.

After a partial write failure is encountered, when the store next resumes processing then the store should first reprocess records which it has already processed previously (if EOS is not enabled), including the one during which the failure occurred, before processing any newly seen records. In this case, we would only expect a single record to be truncated (and to be truncated in full) as part of the recovery process -- as you've outlined above. The recovery code is more general, though, in that it also supports truncating multiple records and/or partial records, in case my above understanding about the types of failures which can occur is not accurate or changes in the future.

Reply:
> The recovery code is more general, though, in that it also supports truncating multiple records and/or partial records, in case my above understanding about the types of failures which can occur is not accurate or changes in the future.

Yes, that is what I figured -- wondering if it would be safe to make it more generic or not? -- The order in which we process records is not 100% deterministic (we could have some `s1.merge(s2).toTable()` for example and read from two partitions).

Or in the "worst" case, somebody might write a non-deterministic custom `Processor`...

Just want to double check your thoughts on this.

Reply:
I actually think the examples you gave are great examples for why we _need_ this logic to be more generic -- if those are "valid" cases which could be encountered during processing/re-processing, then we need the store to be able to properly handle them after encountering a failure. 

The cleanup logic is safe because we the only type of failure we can have is duplicated data, i.e., two segments (or one segment and the latest value store) contain overlapping records. When this happens, we know it is safe to truncate from the older segment because the data being truncated is also present in the newer segment.

Reply:
Updated the comment in the latest commit.